### PR TITLE
collect: add range opertor and date compare. resolves #681

### PR DESCRIPTION
**Pull request by AlicVB**
you can use operators with date-time
range operator use the [ ; ] notation and can be used with all numeric and date values

### CODE REVIEW

File: src/common/collection.c [hunk 1]
```diff
@@ -535,26 +535,132 @@ GList *dt_collection_get_selected(const dt_collection_t *collection, int limit)
...
+    *number1 = g_match_info_fetch(match_info, 1);
+    *number2 = g_match_info_fetch(match_info, 2);
+    *operator= g_strdup("[]");
+    g_match_info_free(match_info);
+    g_regex_unref(regex);
+    return;
+  }
+
+  // and we test the classic comparaison operators
+  regex = g_regex_new("^\\s*(=|<|>|<=|>=|<>)?\\s*([0-9]+\\.?[0-9]*)\\s*$", 0, 0, NULL);
```
Comment (LebedevRI) [inline src/common/collection.c]: `g_regex_unref(regex);`
before this line

Comment (LebedevRI) [inline src/common/collection.c]: and `g_match_info_free(match_info);`

Comment (AlicVB) [inline src/common/collection.c]: oops... and same in the next fct...
fixed

File: src/common/collection.c [hunk 2]
```diff
@@ -535,26 +535,138 @@ GList *dt_collection_get_selected(const dt_collection_t *collection, int limit)
...
+    *number1 = g_match_info_fetch(match_info, 2);
+    // we fill the time part if it's not set
+    if(match_count == 4)
+    {
+      gchar *nb_time = g_match_info_fetch(match_info, 3);
+      *number1 = dt_util_dstrcat(*number1, "%s", nb_time);
+      g_free(nb_time);
+    }
+    else if(*operator&& strcmp(*operator, ">") == 0)
+      *number1 = dt_util_dstrcat(*number1, " 23:59:59");
```
Comment (houz) [inline src/common/collection.c]: Seconds can go to 60. Not sure if that is important here.

**Comment by LebedevRI** [comment]
So
1. It only accepts date in format `2015:01:01`, not `2015:1:01`, not `2015:01:1`, not `2015:1:1`.
2. No tooltip of which format one should actually input.
3. With some range set, e.g. `[2012:1:1; 2013:12:31]`, some images are shown in lt, but no dates are shown in collect. I expected all the dates that fit current filter to be listed there.

Now, for ISO:
1. Filter: `[0; 800]`. no images displayed in lt, all the iso's are still listed in collect images
2. Filter: `[0;800]` (just removed a space - ` `) - all good, like i expected - images displayed in lt, matching iso's are showed up in the list. *It should be much more tolerable to spaces :)*

File: src/common/collection.c (hunk 2 above)
Comment (AlicVB) [inline src/common/collection.c]: oh... leap seconds, even 61 for double leap seconds.
but I don't think we should care, the probability is really quite low, and I'm not sure how systems will handle 60/61 values in "normal" cases (and not sure that all systems are consistent...)

Comment (AlicVB) [inline src/common/collection.c]: and sqlite seems to accept values only in [0-59] : http://www.sqlite.org/lang_datefunc.html
not that we care atm, as anyway we only do string comparisons ;)

**Comment by AlicVB** [comment]
fixed "space tolerance" and tooltips for operators and date format

> 1. It only accepts date in format 2015:01:01, not 2015:1:01, not 2015:01:1, not 2015:1:1
Sadly, date are store as strings in db, so we have not really the choice (apart to make complex date readers) maybe one day we'll have to change date in db...

> 3. With some range set, e.g. [2012:01:01; 2013:12:31], some images are shown in lt, but no dates are shown in collect
This can be added in the future, but I have plan to change the gui for date-time and numeric values

File: src/common/collection.c (hunk 2 above)
Comment (houz) [inline src/common/collection.c]: That is interesting, as `man date` says that seconds can go to 60. But you are probably right and we can ignore it.

**Pull request merged**